pkg/util/debugutil: introduce debug.Stack() wrapper

arjunmahishi opened the pull request:
Until now, all stack traces were treated as untyped strings and hence redacted. Stack traces don't need to be redacted. Because they don't contain any PII. This PR fixes that with the following commits:

1. Introduce a new function `Stack()` in the `debugutil` package as a replacement for the standard `debug.Stack()`. This new function wraps the stack trace contents in `redact.Safe()` to make sure that it does not get redacted.

2. Replace the all the current uses of `debug.Stack()` across the codebase with `debugutil.Stack()`

3. Add a linter rule to block future uses of `debug.Stack()`. The linter error will nudge users to use `debugutil.Stack()` instead.

Part of: CRDB-15292
Epic: CRDB-37533
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136288)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
Can you split the 3 bullet points into separate commits? Easier to review search/replace operations when they're confined to a commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136288#-:-OChtj-N6w3r4pdvDzSK:b-zdw8i5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @itsbilal, @kyle-a-wong, @nameisbhaskar, and @vidit-bhat)

___
*[`pkg/util/debugutil/debugutil.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136288#-OChuPOB-yxMy_t99vMR:-OChuPOB-yxMy_t99vMS:boywxfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/539280454d57d57cc71388ff9c0e4e8a30b84785/pkg/util/debugutil/debugutil.go#L53)):*
> ```Go
> // unnecessary redaction.
> func Stack() interfaces.SafeValue {
> 	return redact.Safe(string(debug.Stack()))
> ```

I believe `string` will allocate a copy of the stack byte slice.
Can you do this via type alias and avoid an allocation?

```
type SafeStack []byte

func (SafeStack) SafeValue() {}

SafeStack(debug.Stack())
```

Alternatively, you can use `unsafe.String` if the above doesn't work.


<!-- Sent from Reviewable.io -->

itsbilal commented on pkg/util/debugutil/debugutil.go:
```diff
@@ -43,3 +46,9 @@ func init() {
+
+// Stack wraps the output of debug.Stack() with redact.Safe() to avoid
+// unnecessary redaction.
+func Stack() interfaces.SafeValue {
```
Given we've had issues in the past with high system CPU usage due to excessive use of debug.Stack in error handling pathways, we should probably add a warning comment here to the effect of "Calling debug.Stack grabs system-level locks and could cause high system CPU usage resulting in the Go runtime to lock up if called too frequently, even if called only in error-handling pathways. Use sporadically and only when necessary."

arjunmahishi replied:
Done
<!-- Reviewable comment -OCm-rTDAK-Y91DKqt3e:bb6un1j -->
<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
mostly a transit-by review as I got tagged

arjunmahishi committed:
pkg/*: replace uses of debug.Stack() with debugutil.Stack()

A previous commit introduced the debugutil.Stack() function which
wraps the debug.Stack() output in "SafeStack" type to make sure that it
does not get redacted.

This commit replaces all the current uses of debug.Stack() with
debugutil.Stack(). With one exception - there is still a call to
debug.Stack() in pkg/util/log/clog.go. Changing this is not that
straightforward. But it's ok keep it for now because this does not go
through the redaction flow.

Part of: CRDB-15292
Epic: CRDB-37533
Release note: None

arjunmahishi committed:
pkg/util/debugutil: introduce debug.Stack() wrapper

Until now, all stack traces were treated as untyped strings and hence
redacted. Stack traces don't need to be redacted. Because they don't
contain any PII.

This commit introduces a new function, Stack(), in the debugutil package
as a replacement for the standard debug.Stack(). The new function wraps
the stack trace contents in a []byte-aliased type "SafeStack" that
implements the SafeValue interface, ensuring it is not redacted. This
function can be used as a drop-in replacement for debug.Stack().

Part of: CRDB-15292
Epic: CRDB-37533
Release note: None

arjunmahishi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136288)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @dhartunian, @itsbilal, @kyle-a-wong, @nameisbhaskar, and @vidit-bhat)

___
*[`pkg/util/debugutil/debugutil.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136288#-OChuPOB-yxMy_t99vMR:-OCm-saYDol6J9dtYTu9:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/539280454d57d57cc71388ff9c0e4e8a30b84785/pkg/util/debugutil/debugutil.go#L53)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I believe `string` will allocate a copy of the stack byte slice.
Can you do this via type alias and avoid an allocation?

```
type SafeStack []byte

func (SafeStack) SafeValue() {}

SafeStack(debug.Stack())
```

Alternatively, you can use `unsafe.String` if the above doesn't work.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
Thanks for splitting this up. LGTM, just some minor questions.

> Reviewed 3 of 43 files at r1, 41 of 41 files at r2, 39 of 39 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136288#-:-ODEpANh2iQiC7RiVeJG:b-u1mvxi)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @itsbilal, @kyle-a-wong, @nameisbhaskar, and @vidit-bhat)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go` line 871 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136288#-ODEoSAM4alSCV8WyJuW:-ODEnbAF4WOe4le2kZRi:b-mzuiye) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccd63f7d7c84cdcd7a37616f9becf050f5d954e5/pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go#L871)):*
> ```Go
> // panic message.
> func logPanicToErr(l *logger.Logger, r interface{}) error {
> 	l.Printf("panic stack trace:\n%s", debugutil.Stack())
> ```

Do you know why the `string()` wrapper was used earlier? I think `%s` works with `[]byte` directly.
___
*[`pkg/testutils/lint/lint_test.go` line 2895 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/136288#-ODEovfX4e68Ep1UrUXb:-ODEovfX4e68Ep1UrUXc:bb6p9jb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b9574d2f209f295f894a51d42d26f864cd7fb08/pkg/testutils/lint/lint_test.go#L2895)):*
> ```Go
> 			// through the redaction process. And changing it to debugutil.Stack()
> 			// would require a lot more changes
> 			":!util/log/clog.go",
> ```

Can you elaborate on what blocks the change from happening?

Separately, is it possible to ignore this lint using an inline comment? It would be nice to have the explanation next to the usage. Not a huge deal though.


<!-- Sent from Reviewable.io -->

arjunmahishi committed:
pkg/testutils/lint: add a linter rule to block call to debug.Stack()

Some of the previous commits addressed the unnecessary redaction of
stack traces by replacing uses of debug.Stack() with debugutil.Stack().
The new function ensures proper handling of redaction.

This commit future-proofs the solution by adding a linter rule to block
future uses of debug.Stack(). The linter error will prompt users to use
debugutil.Stack() instead.

Part of: CRDB-15292
Epic: CRDB-37533
Release note: None

arjunmahishi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136288)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @dhartunian, @itsbilal, @kyle-a-wong, @nameisbhaskar, and @vidit-bhat)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go` line 871 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136288#-ODEoSAM4alSCV8WyJuW:-ODGK4-70VK0YL0URX7I:b-bc6syr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccd63f7d7c84cdcd7a37616f9becf050f5d954e5/pkg/cmd/roachtest/roachtestutil/mixedversion/runner.go#L871)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Do you know why the `string()` wrapper was used earlier? I think `%s` works with `[]byte` directly.
</blockquote></details>

Ya. I wondered the same thing. `[]byte` does work with `%s`. I doubt that there was any reason. I saw this in other places too. They probably assumed that `[]byte` would not work with `%s`.

I was discussing this with @aa-joshi the other day. We need to add some kind of a linting rule to block using `string()` on the args of logging functions. Because it dismisses the redaction context by making it a plain string
___
*[`pkg/testutils/lint/lint_test.go` line 2895 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/136288#-ODEovfX4e68Ep1UrUXb:-ODGSlTK3gUG8z8UmfbF:bsnr6xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b9574d2f209f295f894a51d42d26f864cd7fb08/pkg/testutils/lint/lint_test.go#L2895)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Can you elaborate on what blocks the change from happening?

Separately, is it possible to ignore this lint using an inline comment? It would be nice to have the explanation next to the usage. Not a huge deal though.
</blockquote></details>

Ohh good that you pointed this out!

In my previous implementation, `debugutil.Stack()` was returning `string`. So, it was hard to refactor this part of the code. But now that we are returning the type `SafeStack`, which is an alias to []byte, everything just work as it is. 

I've removed this exception in the linter and made the change in the clog.go file (change `debug.Stack()` to `debugutil.Stack()`)


<!-- Sent from Reviewable.io -->

arjunmahishi commented:
bors r=dhartunian
